core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -1110,6 +1110,21 @@ private RecordCursorFactory generateSelectChoose(QueryModel model, SqlExecutionC
     private RecordCursorFactory generateSelectDistinct(QueryModel model, SqlExecutionContext executionContext) throws SqlException {
         final RecordCursorFactory factory = generateSubQuery(model, executionContext);
         try {
+            if (model.getColumns().size() == 1 && model.getNestedModel() != null && model.getNestedModel().getNestedModel() != null && model.getNestedModel().getNestedModel().getTableName() != null) {
+                ExpressionNode tableName = model.getNestedModel().getNestedModel().getTableName();
+                TableReader reader = engine.getReader(executionContext.getCairoSecurityContext(), tableName.token);
+                CharSequence columnName = model.getColumnNames().get(0);
+                int columnIndex = reader.getMetadata().getColumnIndex(columnName);
+                int columnType = reader.getMetadata().getColumnType(columnIndex);
+                if (columnType == ColumnType.SYMBOL) {
+                    return new DistinctSymbolRecordCursorFactory(
+                            factory,
+                            entityColumnFilter,
+                            reader,
```

Comment:
Reader must not be held by factory. This is a resource hungry instance, which are shared through the engine. Factory must hold on to the CairoEngine instance instead and table name.